src/v/cluster/partition.cc

```diff
@@ -60,13 +61,40 @@ ss::future<result<raft::replicate_result>> partition::replicate(
     return _raft->replicate(std::move(r), opts);
 }
 
+raft::replicate_stages partition::replicate_in_stages(
+  model::record_batch_reader&& r, raft::replicate_options opts) {
+    return _raft->replicate_in_stages(std::move(r), opts);
+}
+
 ss::future<result<raft::replicate_result>> partition::replicate(
   model::term_id term,
   model::record_batch_reader&& r,
   raft::replicate_options opts) {
     return _raft->replicate(term, std::move(r), opts);
 }
 
+raft::replicate_stages partition::replicate_in_stages(
+  model::batch_identity bid,
+  model::record_batch_reader&& r,
+  raft::replicate_options opts) {
+    if (bid.is_transactional || bid.has_idempotent()) {
+        ss::promise<> p;
+        auto f = p.get_future();
+        auto replicate_finished
+          = _rm_stm->replicate(bid, std::move(r), opts)
+              .then(
+                [p = std::move(p)](result<raft::replicate_result> res) mutable {
+                    p.set_value();
+                    return res;
```

Comment:
@mmaslankaprv @rystsov this looks like a broken promise `p` if replicate fails? other wrappers in this PR use `then_wrapped` and handle the failure cases.